admin/javascript/src/components/form/index.js

```diff
@@ -0,0 +1,65 @@
+import React from 'react';
+import SilverStripeComponent from 'silverstripe-component';
+import FormActionComponent from 'components/form-action/index';
+
+class FormComponent extends SilverStripeComponent {
+
+  /**
+   * Gets the components responsible for perfoming actions on the form.
+   * For example form submission.
+   *
+   * @return array|null
+   */
+  getFormActionComponents() {
+    return this.props.actions.map((action) =>
+      <FormActionComponent {...action} />
+    );
+  }
+
+  render() {
+    const attr = this.props.attributes;
+    const fields = this.props.mapFieldsToComponents(this.props.fields);
```

Comment:
Since there's only a limited number of props passed in here which can lead to state tree mutations, I think that's acceptable in this case. Also, don't really know how to cache this outside of `render()` - against a nested object identity of `this.props.fields`? A mutation in that prop is the most likely bit to trigger the re-render in the first place, so will already have the same cache logic.